lemmy_apub/src/activities/receive/mod.rs

```diff
@@ -121,3 +123,49 @@ pub(crate) async fn find_by_id(
 
   return Err(NotFound.into());
 }
+
+pub(crate) fn verify_activity_domains_valid<T, Kind>(
+  activity: &T,
+  expected_domain: Url,
+  object_domain_must_match: bool,
+) -> Result<(), LemmyError>
+where
+  T: AsBase<Kind> + ActorAndObjectRef,
+{
+  let actor = activity
+    .actor()?
+    .to_owned()
+    .single_xsd_any_uri()
+    .context(location_info!())?;
+  let expected_domain = expected_domain.domain().context(location_info!())?;
+
+  activity.id(expected_domain)?;
+  if actor.domain() != Some(expected_domain) {
+    return Err(DomainError.into());
+  }
+
+  let object_id = match activity.object().to_owned().single_xsd_any_uri() {
+    // object is just an ID
+    Some(id) => id,
+    // object is something like an activity, a comment or a post
+    None => activity
+      .object()
+      .to_owned()
+      .one()
+      .context(location_info!())?
+      .id()
+      .context(location_info!())?
+      .to_owned(),
+  };
+
+  if !object_domain_must_match {
+    check_is_apub_id_valid(&object_id)?;
+    return Ok(());
+  }
+
+  if object_id.domain() != Some(expected_domain) {
+    return Err(DomainError.into());
+  }
+
+  Ok(())
+}
```

Comment:
So you've got an actor domain check, an object domain check, and an apub_id_is_valid (IE blocks and such) check.

What's `object_domain_must_match` about? The name doesn't make sense to me, shouldn't it be wrapped around like this: 

```
  if object_domain_must_match && (object_id.domain() != Some(expected_domain)) {
    return Err(DomainError.into());
  }
```

and not this?

```
  if !object_domain_must_match {
    check_is_apub_id_valid(&object_id)?;
    return Ok(());
  }
```

Seems weird to have a boolean called object_domain_must_match, but then run the object domain check no matter what anyway.

I think we should always run the check_is_apub_id_valid no matter what, but the inner actor and object check should be conditional, since `announce` should be allowed to have actors from a different domain.

Reply:
You're right, the variable names are confusing. `expected_domain` should rather be called `actor_id`, because that one already gets checked by the inbox, including http signature, so we know its valid.

As for `object_domain_must_match`, I thought it would be good to check the inner object against block/allowlist because there we dont know anything about it. Most likely that check isnt needed though, as the inner object would be checked elsewhere. So I should probably change it to the code you suggested.

Reply:
Done, also I removed the check of the actor domain in this method (it would just be compared against itself).